src/ILCompiler.Compiler/src/CppCodeGen/CppWriter.cs

```diff
@@ -729,192 +733,204 @@ private String GetCodeForVirtualMethod(MethodDesc method, int slot)
             return sb.ToString();
         }
 
-        private void AppendVirtualSlots(CppGenerationBuffer sb, TypeDesc implType, TypeDesc declType)
+        private String GetCodeForNode(DependencyNode node, NodeFactory factory)
         {
-            var baseType = declType.BaseType;
-            if (baseType != null)
-                AppendVirtualSlots(sb, implType, baseType);
+            // virtual slots
+            var objectNode = node as ObjectNode;
+            var nodeData = objectNode.GetData(factory, false);
+
+            CppGenerationBuffer nodeCode = new CppGenerationBuffer();
+
+            /* Create list of byte data. Used to divide contents between reloc and byte data
+          * First val - isReloc
+          * Second val - size of byte data if first value of tuple is false
+          */
+
+            List<Tuple<bool, int>> nodeDataDivs = new List<Tuple<bool, int>>();
+            byte[] actualData = new byte[nodeData.Data.Length];
+            Relocation[] relocs = nodeData.Relocs;
+
+            int nextRelocOffset = -1;
+            int nextRelocIndex = -1;
+            int lastByteIndex = 0;
 
-            IReadOnlyList<MethodDesc> virtualSlots = _compilation.NodeFactory.VTable(declType).Slots;
-            for (int i = 0; i < virtualSlots.Count; i++)
+            if (relocs.Length > 0)
             {
-                MethodDesc declMethod = virtualSlots[i];
-                MethodDesc implMethod = implType.GetClosestDefType().FindVirtualFunctionTargetMethodOnObjectType(declMethod);
+                nextRelocOffset = relocs[0].Offset;
+                nextRelocIndex = 0;
+            }
 
-                sb.AppendLine();
-                if (implMethod.IsAbstract)
+            int i = 0;
+
+            int j = 0;
+            CppGenerationBuffer nodeDataDecl = new CppGenerationBuffer();
+            int offset = 0;
+            if (node is ISymbolNode)
+            {
+                offset = (node as ISymbolNode).Offset;
+                i = offset;
+                lastByteIndex = offset;
+            }
+            while (i < nodeData.Data.Length)
+            {
+                if (i == nextRelocOffset)
                 {
-                    sb.Append("NULL,");
+                    Relocation reloc = relocs[nextRelocIndex];
+
+                    int size;
+                    switch (reloc.RelocType)
+                    {
+                        case RelocType.IMAGE_REL_BASED_DIR64:
+                            size = 8;
+                            break;
+                        case RelocType.IMAGE_REL_BASED_REL32:
```

Comment:
I do not think `IMAGE_REL_BASED_REL32` relocs are going to work for CppCodeGen.

I think you should just assert here that you are only getting pointer reloc here, something like:
`
Debug.Assert(reloc.RelocType == (PointerSize == 8) ? RelocType.IMAGE_REL_BASED_DIR64 : RelocType.IMAGE_REL_BASED_HIGHLOW);
`